core/trino-main/src/main/java/io/trino/sql/planner/PlanOptimizers.java

```diff
@@ -644,6 +654,10 @@ public PlanOptimizers(
                 .add(new PushDistinctLimitIntoTableScan(plannerContext))
                 .add(new PushTopNIntoTableScan(metadata))
                 .add(new RewriteTableFunctionToTableScan(plannerContext)) // must run after ImplementTableFunctionSource
+                .add(new MaterializeFilteredTableScan(plannerContext, splitManager, pageSourceManager, nodeManager, executor))
+                .add(new MaterializeTableScan(plannerContext, splitManager, pageSourceManager, nodeManager, executor))
```

Comment:
This might be too early in the planning process for this optimization. Ideally, we want to delay this until after as much predicate and projection pushdown as possible has happened.
I think we should at least do it after or near `io.trino.sql.planner.optimizations.MetadataQueryOptimizer` runs as that is a similar optimization as this one and we probably want that one to run that one first.

Reply:
I have NO IDEA where to put this optimization. I just looked around and guessed. If you can tell me exactly where to put it, that would be great. Or if you want to take over this PR, I'm happy for that too..